src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentBase.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
+{
+    public abstract class VirtualDocumentBase<T> : VirtualDocument where T : VirtualDocumentSnapshot
+    {
+        private T _currentSnapshot;
+        private long? _hostDocumentSyncVersion;
+
+        protected VirtualDocumentBase(Uri uri, ITextBuffer textBuffer)
+        {
+            if (uri is null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            if (textBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            Uri = uri;
+            TextBuffer = textBuffer;
+            _currentSnapshot = GetUpdatedSnapshot();
+        }
+
+        public override Uri Uri { get; }
+
+        public override ITextBuffer TextBuffer { get; }
+
+        public override long? HostDocumentSyncVersion => _hostDocumentSyncVersion;
+
+        public override VirtualDocumentSnapshot CurrentSnapshot => _currentSnapshot;
+
+        public override VirtualDocumentSnapshot Update(IReadOnlyList<ITextChange> changes, long hostDocumentVersion)
+        {
+            if (changes is null)
+            {
+                throw new ArgumentNullException(nameof(changes));
+            }
+
+            _hostDocumentSyncVersion = hostDocumentVersion;
+            TextBuffer.SetHostDocumentSyncVersion(_hostDocumentSyncVersion.Value);
+
+            if (changes.Count == 0)
+            {
+                // Even though nothing changed here, we want the synchronizer to be aware of the host document version change.
+                // So, let's make an empty edit to invoke the text buffer Changed events.
+                TextBuffer.MakeEmptyEdit();
+
+                _currentSnapshot = GetUpdatedSnapshot();
+                return _currentSnapshot;
+            }
+
+            using var edit = TextBuffer.CreateEdit(EditOptions.None, reiteratedVersionNumber: null, InviolableEditTag.Instance);
+            for (var i = 0; i < changes.Count; i++)
```

Comment:
This code seems unnecessarilly complex. Couldn't it just be:

            for (var i = 0; i < changes.Count; i++)
            {
                var change = changes[i];

                edit.Replace(change.OldSpan.Start, change.OldSpan.Length, change.NewText);
            }

Reply:
Yeah, I left the original code there, not sure if there is / was a reason for complexity. I'll go with your suggestion Todd unless Taylor tells me why I shouldn't. 

---
In reply to: [450338503](https://github.com/dotnet/aspnetcore-tooling/pull/2174#discussion_r450338503) [](ancestors = 450338503)

Reply:
> Yeah, I left the original code there, not sure if there is / was a reason for complexity. I'll go with your suggestion Todd unless Taylor tells me why I shouldn't.

I figured the TextEdit applier does different things for each so in my mind I was taking the "most optimal" route by using the appropriate method.

Reply:
The Insert/Delete methods just call Replace under the hood:

[https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FText%2FImpl%2FTextModel%2FBaseBuffer.cs&version=GBmaster&line=216&lineEnd=217&lineStartColumn=1&lineEndColumn=1&lineStyle=plain](https://devdiv.visualstudio.com/DevDiv/_git/VS-Platform?path=%2Fsrc%2FEditor%2FText%2FImpl%2FTextModel%2FBaseBuffer.cs&version=GBmaster&line=216&lineEnd=217&lineStartColumn=1&lineEndColumn=1&lineStyle=plain)

---
In reply to: [450397135](https://github.com/dotnet/aspnetcore-tooling/pull/2174#discussion_r450397135) [](ancestors = 450397135)

Reply:
wow! Alright, looks like I was just doing extra unnecessary work 😆